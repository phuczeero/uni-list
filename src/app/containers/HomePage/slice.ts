import { PayloadAction } from '@reduxjs/toolkit';
import { University } from 'types/University';
import { createSlice } from 'utils/@reduxjs/toolkit';
import { ContainerState } from './types';

// The initial state of the HomePage container
export const initialState: ContainerState = {
  searchTerm: '',
  universities: [],
  loading: false,
  error: null,
};

const homePageSlice = createSlice({
  name: 'homePage',
  initialState,
  reducers: {
    loadUniversities(state, action: PayloadAction<string>) {
      state.searchTerm = action.payload;
      state.loading = true;
    },
    loadSuccess(state, action: PayloadAction<University[]>) {
      state.loading = false;
      state.universities = action.payload;
    },
    loadError(state, action: PayloadAction<Error>) {
      state.error = action.payload;
      state.loading = false;
    },
    saveToJSON() {},
  },
});

export const {
  actions: homePageActions,
  reducer,
  name: sliceKey,
} = homePageSlice;
