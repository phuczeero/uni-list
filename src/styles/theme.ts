const fonts = {
  body:
    '-apple-system, BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji"',
};
const fontSizes = [12, 14, 16, 20, 24, 32, 48, 64, 72];
const space = [0, 4, 8, 16, 32, 64, 128, 256, 512];
const radii = {
  default: 6,
  circle: 99999,
};
const colors = {
  text: '#212529',
  muted: '#dee2e6',
  primary: '#007bff',
  danger: '#dc3545',
  blues: {
    light: '#dcedff',
    dark: '#004896',
  },
};
const sizes = {
  navbar: 64,
  spinner: 32,
};
const shadows = {
  card: '0 0 4px rgba(0, 0, 0, .125)',
};
const breakpoints = ['576px', '768px', '992px', '1200px'];

export const theme = {
  fonts,
  fontSizes,
  space,
  radii,
  colors,
  sizes,
  shadows,
  breakpoints,
};

export type ThemeType = typeof theme;
