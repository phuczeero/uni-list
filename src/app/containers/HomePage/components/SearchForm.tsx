import React, { FormEvent, useRef, useState } from 'react';
import { useDispatch } from 'react-redux';
import { homePageActions } from '../slice';
import { SearchField } from 'app/components/SearchField';
import { ErrorMessage } from './ErrorMessage';

const MIN_LENGTH = 3;

interface Props {}

export function SearchForm(props: Props) {
  const dispatch = useDispatch();
  const ref = useRef<HTMLInputElement>(null!);
  const [invalid, setInvalid] = useState(false);

  const onSubmit = (e: FormEvent) => {
    e.preventDefault();
    const searchTerm = ref.current.value;
    const invalid = searchTerm.length < MIN_LENGTH;
    setInvalid(invalid);
    if (!invalid) {
      dispatch(homePageActions.loadUniversities(searchTerm));
    }
  };

  const onClose = () => {
    setInvalid(false);
    ref.current.focus();
  };

  const errorMessage = (
    <ErrorMessage
      message="Please enter at least 3 characters to search."
      onClose={onClose}
    />
  );

  return (
    <form onSubmit={onSubmit}>
      <SearchField ref={ref} placeholder="Search universities" />
      {invalid && errorMessage}
    </form>
  );
}
