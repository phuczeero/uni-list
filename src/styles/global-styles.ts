import { createGlobalStyle } from 'styled-components';
import { ThemeType } from './theme';

export const GlobalStyle = createGlobalStyle<{ theme: ThemeType }>`
  * {
    font-family: ${p => p.theme.fonts.body};
    color: ${p => p.theme.colors.text}
  }
`;
