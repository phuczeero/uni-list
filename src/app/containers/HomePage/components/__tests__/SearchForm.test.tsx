import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { SearchForm } from '../SearchForm';
import { ThemeProvider } from 'styled-components';
import { theme } from 'styles/theme';
import { Provider } from 'react-redux';
import { configureAppStore } from 'store/configureStore';
import { homePageActions } from '../../slice';

const renderSearchForm = store =>
  render(
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <SearchForm />,
      </ThemeProvider>
    </Provider>,
  );

describe('<SearchForm />', () => {
  let store: ReturnType<typeof configureAppStore>;
  beforeEach(() => {
    store = configureAppStore();
  });

  it('should render and match the snapshot', () => {
    const component = renderSearchForm(store);
    expect(component.container.firstChild).toMatchSnapshot();
  });

  it('should handle valid submit', () => {
    const mock = jest.fn();
    const mockStore = { ...store, dispatch: mock };
    const component = renderSearchForm(mockStore);
    const input = component.getByPlaceholderText('Search universities');
    const form = component.container.querySelector('form') as HTMLFormElement;
    const searchTerm = 'test';
    fireEvent.change(input, { target: { value: searchTerm } });
    fireEvent.submit(form);
    expect(mock).toBeCalledWith(homePageActions.loadUniversities(searchTerm));
  });
});
