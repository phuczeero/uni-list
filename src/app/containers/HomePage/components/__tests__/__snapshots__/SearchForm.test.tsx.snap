// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchForm /> should render and match the snapshot 1`] = `
.c3 {
  outline: none;
  padding: 8px;
  border-radius: 6px;
  border: 2px solid #dee2e6;
}

.c1 {
  position: absolute;
  top: 10px;
  left: 10px;
}

.c0 {
  position: relative;
}

.c0 .c2 {
  padding-left: 32px;
}

<form>
  <div
    class="c0"
  >
    <svg
      class="c1"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
    <input
      class="c2 c3"
      placeholder="Search universities"
    />
  </div>
</form>
`;
